circle.yml

```diff
@@ -379,64 +331,138 @@ jobs:
                   cypressVersion: << parameters.dockerTag >>
-                  imageName: << parameters.dockerName >>:<< parameters.dockerTag >>
+                  imageName: cypress/included:<< parameters.dockerTag >>
 
             - test-included-image:
                   cypressVersion: << parameters.dockerTag >>
-                  imageName: << parameters.dockerName >>:<< parameters.dockerTag >>
-
+                  imageName: cypress/included:<< parameters.dockerTag >>
             - halt-on-branch
-            - docker-push:
-                  imageName: << parameters.dockerName >>:<< parameters.dockerTag >>
 
+    push-images:
+        ## the multi-arch images have been built and tested by previous jobs, so we're safe to blindly build+push
+        machine:
+            image: ubuntu-2004:202111-02
+        parameters:
+            dockerName:
+                type: string
+                description: Image name to build
+            dockerTag:
+                type: string
+                description: Image tag to build, should match Cypress version, like "3.8.1"
+            workingDirectory:
+                type: string
+                description: Directory with the Dockerfile
+            buildxArches:
+                type: string
+                description: List of architectures to pass to `docker buildx --platform`
+        steps:
+            - checkout
+            - halt-if-docker-image-exists:
+                imageName: << parameters.dockerName >>:<< parameters.dockerTag >>
+            - halt-on-branch
+            - run:
+                ## see https://docs.docker.com/desktop/multi-arch/
+                name: build + push multi-arch image to docker hub
+                working_directory: << parameters.workingDirectory >>
+                command: |
+                    ## install QEMU, pulled from https://github.com/docker/setup-qemu-action/blob/4620c11b701e878e3f2b7aa142927d0646857fd9/src/main.ts
+                    ## avoids https://github.com/docker/buildx/issues/499
+                    BINFMT_IMAGE=tonistiigi/binfmt:latest@sha256:85683def11494bc0055d98dd1081e73cb5e70bbc1ae532be3d12cf054e2b11f1
+                    docker pull $BINFMT_IMAGE
+                    docker image inspect $BINFMT_IMAGE
+                    docker run --rm --privileged $BINFMT_IMAGE --install linux/amd64,linux/arm64
+
+                    ## see https://docs.docker.com/desktop/multi-arch/
+                    echo "$DOCKERHUB_PASS" | docker login -u "$DOCKERHUB_USERNAME" --password-stdin
+                    docker buildx create --name builder --use
+                    docker buildx inspect --bootstrap
+                    DOCKER_TAG=<< parameters.dockerName >>:<< parameters.dockerTag >>
+                    docker buildx build --progress plain --platform << parameters.buildxArches >> -t $DOCKER_TAG --push .
+            - run:
+                name: build + push multi-arch image to amazon ecr
```

Comment:
Great work, would have been nice if this change had been broken out. Do you have any testing jobs you can link to that verified this worked as expected?

Reply:
Yeah, I mainly didn't break it out because I didn't want to replicate the work I did for `circle.yml` generation for `buildspec.yml` generation, but I could've just done a feature branch for all of this instead.

Here's my test build on AWS ECR: https://gallery.ecr.aws/s2i6x6o0/flotwig/base

Here is the job that pushed this: https://app.circleci.com/pipelines/github/flotwig/cypress-docker-images/32/workflows/a8a473da-1c09-4651-aa18-be355ad65a3c/jobs/103